### PR TITLE
AG-7122 - Fix legend item click callback

### PR DESCRIPTION
**Pull request by alantreadway**
https://ag-grid.atlassian.net/browse/AG-7122

Fixes the `legendItemClick` event callback for `line` and `scatter` series cases.

Bonus changes:
- Improved Events documentation.
  - Adds legendItemClick example.
  - More concisely documents histogram case.
- Adds improved generic types for chart options, with defaults of any for backwards compatibility.
  - Allows addition and documentation of HistogramBinDatum.

### CODE REVIEW

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      labelKey: "label",
+    },
+  ],
+  ...mixin,
+}
+
+let chart: any = agCharts.AgChart.create(cartesianOptions)
+
+function update(type: string) {
+  chart.destroy()
```
Comment (manapeirov) [inline grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts:103]: nit: missing semicolon

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        options.series = options.series.map((s: any) => {
+          s = { ...s };
+          s.type = type;
+      
+          return s;
+        });
+    }
+  }
+
+  chart = agCharts.AgChart.create(options)
```
Comment (manapeirov) [inline grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts:122]: missing semicolon

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts (hunk 1 above)
Comment (manapeirov) [inline grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts:103]: I checked out this branch and was looking at the example, looks like the framework examples throw an error. I found before as well that trying to switch chart types in one example causes issues for the framework examples. It's fine in js and ts versions

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts (hunk 2 above)
Comment (alantreadway) [reply grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts]: Fixed

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts (hunk 1 above)
Comment (alantreadway) [reply grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts]: Thanks for checking this, I've simplified the example to trade-off compatibility for coverage.

**Comment by alantreadway** [comment]
@manapeirov PTAL

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts [hunk 3]
```diff
@@ -63,61 +60,6 @@ const cartesianOptions: AgCartesianChartOptions = {
     { type: "number", position: "left" },
   ],
   ...mixin,
```
Comment (manapeirov) [inline grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts:62]: nit: do we need the legend config to be separate still?

**Review by manapeirov — APPROVED** [review]

File: grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts (hunk 3 above)
Comment (alantreadway) [reply grid-packages/ag-grid-docs/documentation/doc-pages/charts-events/examples/legend-item-click-event/main.ts]: Fixed

**Review by manapeirov — APPROVED** [review]

**Pull request merged**